src/compiler/checker.ts

```diff
@@ -5182,7 +5220,14 @@ namespace ts {
         }
 
         function isContextSensitiveFunctionLikeDeclaration(node: FunctionLikeDeclaration) {
-            return !node.typeParameters && node.parameters.length && !forEach(node.parameters, p => p.type);
+            if (compilerOptions.strictThis) {

```

Comment:
I thought that in loose-this mode, a function with all arg types specified and no `this` is not contextually typed. But thinking about it a second time, I don't think that's true:

```ts
interface I {
  y: number;
  m(this: I, x: number);
}
let o: I = {
  m(x: number) {
    return x + this./**/
  }
}
```

It would in fact be nice if we would contextually type `this` so that we could offer `y` as a completion. I'll get rid of everything except the updated check.